### ISSUE TITLE
ViewableData/SSViewer caching issues, behavioural change (3.2)

### ISSUE THREAD

**Issue by kinglozzer** [silverstripe/silverstripe-framework#4626]
Firstly, in the example below, I believe “Running getter” should only be logged once when it is in fact logged 3 times:

```html
<!-- Page.ss -->
<% if $Test %>
	<% with $Test %>
		$Message
	<% end_with %>
	
	$Test.Message
<% end_if %>
```
```php
// Page_Controller
public function Test() {
	Debug::log('Running getter');
	return ArrayData::create(array('Message' => 'Hi'));
}
```

Running an `<% if $Test %>` **does** cache the value. However, using either `<% with $Test %>` or `$Test.Message` **does not** cache the value - both are translated internally to `obj('Test', null, true)`. This can be fixed by adding a fourth parameter in the template parser to enable caching (`obj('Test', null, true, true)`), but I wanted to log an issue to get some extra eyes on this before I create a PR.

---

The actual behavioural change between 3.1 and 3.2 is caused by this issue and #3985. Before that PR was merged, `ViewableData` would read data from the cache _even if the `$cache` parameter was false_. So things went in this order:

- `<% if $Foo %>` resulted in a cached call, the value of `$Foo` was added to the cache;
- `<% with $Foo %>` resulted in a non-cached call, but `SSViewer` saw the data in the cache and returned it anyway.

After #3895, `SSViewer` will avoid the cache entirely if `$cache` is false - even if the value already exists in the cache. [This gist](https://gist.github.com/kinglozzer/7e21295459bc9fd7dc47) contains my sample code that used to work in 3.1, but no longer works in 3.2.

**Comment by kinglozzer** [comment]
Looking at this again this morning, my initial attempt won’t cut it. I think the main problem is that value lookups assume the method signature is identical for `XML_val()`, `hasValue()` and `obj()` - but `obj()` has an extra argument (`$forceReturnedObject`) before `$cache` unlike the others.

I’ve got a working solution that relies on `preg_replace()` to manually add a fourth parameter to `obj()` lookups, I’ll see if I can tidy it up and capture this in a test later.

Sorry that this is a bit tricky to explain... hopefully a test case will at least demonstrate the behavioural change clearly!

**Comment by dhensby** [comment]
nice work spotting this

**Comment by tractorcow** [comment]
I've made some comments on your pull request. My feeling is that maybe we should fix this alongside @dhensby 's default_cast fix in 4.0.

**Comment by kinglozzer** [comment]
This has been resolved in master by https://github.com/silverstripe/silverstripe-framework/pull/5653. The original problem described in this issue is now fixed, and the tests in https://github.com/silverstripe/silverstripe-framework/pull/4628 pass
